text/0006-shipyard-next-generation.md

```diff
@@ -0,0 +1,317 @@
+
+Please read this KEP in combination with the KEP: [First version of a Keptn uniform](https://github.com/keptn/enhancement-proposals/pull/7); especially regarding the eventing part.
+
+## Explanation
+
+:man:/:blonde_woman: The target persona of this KEP is an *SRE*, who is responsible for processes and workflows to automate application delivery or remediation operations. 
+
+Before getting started, a brief recap of the important terms and definitions used in this KEP:
+
+- **Project stage:** A project stage (or just stage) defines a logical space in an environment, which has a dedicated purpose for an application in a continuous delivery process. The project stages are ordered, meaning that a stage has at least a previous and/or next stage. 
+
+- **Shipyard:** A shipyard is the declarative means to divide an environment into project stages and to specify task sequences for each stage.
+
+- **Sequence:** A sequence declares a set of tasks, which are triggered by an external event (aka. domain event, e.g.: artifact built → deploy artifact, a problem occurred → execute remediation, etc.). A sequence executes the tasks in the given order. 
+
+- **Task:** A task is the smallest executable unit of a sequence. A task is kicked-off by a *triggered* event and has a life cycle of a *start* and *finished* event. 
+
+*Current situation and Problem(s):* The current shipyard specification allows dividing an environment into project stages. In each stage, two ways for deploying (direct, blue/green) and testing (functional/performance) an artifact are provided. Besides, an SRE can specify whether a project stage is capable of handling problems by remediation actions. This specification has some limitations: 
+
+- Certain tasks are defined implicitly without giving an SRE the possibility to remove/move the task, e.g.:
+
+  - The evaluation always happens after a test execution even though the SRE already knows that no meaningful metric values are available (e.g., functional tests don't provide performance data). 
+
+  - The promotion of a deployed/released artifact happens automatically with no possibility to intervene.
+
+- It is not possible to add a task to a sequence, e.g., a task that takes care of pulling approval from a person in charge. 
+
+- The trigger (domain event) of a task sequence can not be set. 
+
+*Solution:* A revision of the shipyard specification v0.1.2 addresses the above-mentioned problems and gives SREs more flexibility in declaring processes. 
+
+## Internal details
+
+### Specification 
+
+The KEP proposes a new version for the shipyard specification as explained below. To get started, the following example of a shipyard with one stage is provided: 
+
+```yaml
+---
+version: 0.2.0
+kind: Shipyard
+metadata:
+  name: shipyard-abc
+spec:
+  stages:
+  - name: hardening
+    sequences:
+    - name: artifact-delivery
+      triggers:   
```

Comment:
@johannes-b without explicitly reading the spec, "triggers" can mean both, events that trigger the sequence OR events that are triggered by the sequence. Changing this to e.g., "triggered-by" would be more expressive.

Reply:
I agree, if we only look at the lines 54 to 59, it looks like that artifact-delivery in stage hardening *triggers* a `dev.artifact-delivery.finished`, which does not make any sense.

Looking at the whole file though, it's clear to me that there are stages, a stage has sequences, and a sequence has *triggers*.

I don't like the looks of "triggered-by" in a YAML file, hence I would propose a more generic "on", e.g.:

```yaml
spec:
  stages:
  - name: hardening
    sequences:
    - name: artifact-delivery
      on:  
      - dev.artifact-delivery.finished
      tasks:
      - name: deployment
      - name: test
      - name: evaluation 
      ...
```

Reply:
@johannes-b any update on this? Will be stay with `triggers`?

Reply:
After talking to @jetzlstorfer we also thought about `triggeredByAny` (as the list of sequences is an array which is linked with an `OR` expression)

Reply:
We aggreed on: 
```
on:
  any:
    - sh.keptn.events
    - sh.keptn.events
  
  all:
    - sh.keptn.events
    - sh.keptn.events
```

Reply:
We go with `on`, but will not implement the `any` / `all` property. By default, any applies.